OrbitQt/TutorialOverlay.cpp

```diff
@@ -0,0 +1,423 @@
+  step.cutout_widget = tab->findChild<CutoutWidget*>();
+
+  if (step.cutout_widget) {
+    QRect cutout_rect = step.cutout_widget->geometry();
+    for (auto child : tab->findChildren<QWidget*>()) {
+      if (child != step.cutout_widget && child->parent() == tab) {
+        step.hints.push_back(DeriveHintDescription(cutout_rect, child));
+      }
+    }
+  }
+
+  return step;
+}
+
+void TutorialOverlay::InitAllStepsFromUi() {
+  steps_.clear();
+  for (int i = 0; i < ui_->tabWidget->count(); ++i) {
+    const std::string tab_name = ui_->tabWidget->widget(i)->objectName().toStdString();
+    CHECK(steps_.find(tab_name) == steps_.end());
+    steps_[tab_name] = InitializeStep(i);
+  }
+}
+
+const QRect TutorialOverlay::AbsoluteGeometry(QWidget* widget) const {
+  if (widget->parentWidget() == nullptr || widget->parentWidget() == parentWidget()) {
+    return widget->geometry();
+  }
+
+  QWidget* parent_widget = widget->parentWidget();
+  QRect result = widget->geometry();
+  QRect parent_geo = AbsoluteGeometry(parent_widget);
+
+  auto width = result.width();
+  auto height = result.height();
+
+  result.setX(result.x() + parent_geo.x());
+  result.setY(result.y() + parent_geo.y());
+
+  result.setWidth(width);
+  result.setHeight(height);
+
+  return result;
+}
+
+TutorialOverlay::~TutorialOverlay() {}
+
+void TutorialOverlay::StartActiveStep() {
+  const Step& step = GetActiveStep();
+  ui_->tabWidget->setCurrentIndex(step.tab_index);
+
+  if (step.setup.callback_init) {
+    step.setup.callback_init(this, step.name);
+  }
+
+  ui_->btnPrev->setEnabled(!ActiveStepIsFirstInSection());
+  const Section& section = GetActiveSection();
+  const std::string text = absl::StrFormat("%s: %d/%d", section.title,
+                                           (section.active_step - section.step_names.begin()) + 1,
+                                           section.step_names.size());
+  ui_->labelStep->setText(text.c_str());
+
+  if (step.setup.anchor_widget) {
+    step.setup.anchor_widget->installEventFilter(this);
+  }
+  showMaximized();
+  UpdateOverlayLayout();
+}
+
+bool TutorialOverlay::EndActiveStep(bool verify) {
+  if (!HasActiveStep()) {
+    return true;
+  }
+
+  if (verify && !VerifyActiveStep()) {
+    return false;
+  }
+  UnhookActiveStep();
+
+  return true;
+}
+
+void TutorialOverlay::StartSection(const std::string& section) {
+  if (!EndActiveStep(false)) {
+    return;
+  }
+
+  auto it = sections_.find(section);
+  CHECK(it != sections_.end());
+
+  active_section_ = it;
+  it->second.active_step = it->second.step_names.begin();
+  StartActiveStep();
+}
+
+void TutorialOverlay::AddSection(std::string section_name, std::string title,
+                                 std::vector<std::string> step_names) {
+  Section section;
+  for (auto& name : step_names) {
+    // This should just force a check for the existance of the step
+    const StepSetup& step = GetStep(name);
+    UNUSED(step);
+  }
```

Comment:
Lesson learned: If I already think it's a bit ugly and requires a comment to explain why I'm doing this, it will not make it through code review ;-)

You're absolutely right, I changed this in a slightly different way: Added a `StepExists` method that is used here and in `GetStep` to unify this, and added a test for it.